### PR TITLE
HHH-13872 - Make the Java Stream close the underlying ScrollableResultsIterator upon calling a terminal operation

### PR DESCRIPTION
**Pull request by vladmihalcea**
https://hibernate.atlassian.net/browse/HHH-13872

### CODE REVIEW

**Comment by vladmihalcea** [comment]
@NathanQingyangXu @dreab8 @sebersole Would you mind taking a look at this PR?

This PR adds two benefits:

1. It allows us to pass the `closeHandler` when creating a derived `Stream`, `IntStream`, `LongStream`, `DoubleStream` as a result of a `mapToInt`, `mapToLong`, etc method on the original `Stream`
2. We can close the underlying `ResultSet` after calling a terminal operation since the `Stream` is not reusable afterward. If we don't do that, we the `ResultSet` will be open until the `Session` gets closed or the or the currently running transaction is committed or rolled back.

File: hibernate-core/src/main/java/org/hibernate/query/spi/StreamDecorator.java [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+ * The {@link StreamDecorator} wraps a Java {@link Stream} and registers a {@code closeHandler}
+ * which is passed further to any resulting {@link Stream}.
+ *
+ * The goal of the {@link StreamDecorator} is to close the underlying {@link Stream} upon
+ * calling a terminal operation.
+ *
+ * @author Vlad Mihalcea
+ * @since 5.4
+ */
+public class StreamDecorator<R> implements Stream<R> {
```
Comment (NathanQingyangXu) [inline hibernate-core/src/main/java/org/hibernate/query/spi/StreamDecorator.java]: Seems another exciting PR from Vlad. I just wanna raise some awareness of the maintenance cost of the current implementation. So suppose next JVM adds some new method to Stream, do we need to modify the class accordingly to sync up? Seems not a concern for JVM maintains backward compatibility by interface default method?

Other than delegator, Spring's AOP might be a good implementation alternative provided JDK exposes some mean to know whether the current method is 'terminal' one or not.

File: hibernate-core/src/test/java/org/hibernate/test/stream/basic/JpaStreamTest.java [hunk 2]
```diff
@@ -70,12 +79,100 @@ public void testQueryStream() {
...
+			assertFalse( resourceRegistry.hasRegisteredResources() );
+		} );
+	}
+
+	private ResourceRegistry resourceRegistry(Session session) {
+		SharedSessionContractImplementor sharedSessionContractImplementor = (SharedSessionContractImplementor) session;
+		JdbcCoordinator jdbcCoordinator = sharedSessionContractImplementor.getJdbcCoordinator();
+		return jdbcCoordinator.getLogicalConnection().getResourceRegistry();
+	}
+
```
Comment (NathanQingyangXu) [inline hibernate-core/src/test/java/org/hibernate/test/stream/basic/JpaStreamTest.java]: One good edge case for this PR is to verify the following scenario:

- some stream client code has closed stream already (to adhere to best practice, maybe via 'try' block);
- this new feature was added without client knowing and the stream might be closed twice;

It would be excellent to add a testing case to ensure the above common scenario won't throw some exception (not only for now, but for the long future).

File: hibernate-core/src/main/java/org/hibernate/query/spi/StreamDecorator.java [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+import java.util.function.ToDoubleFunction;
+import java.util.function.ToIntFunction;
+import java.util.function.ToLongFunction;
+import java.util.stream.Collector;
+import java.util.stream.DoubleStream;
+import java.util.stream.IntStream;
+import java.util.stream.LongStream;
+import java.util.stream.Stream;
+
+import org.hibernate.Incubating;
```
[the remark this answers is not in the document]
Comment (vladmihalcea) [reply hibernate-core/src/main/java/org/hibernate/query/spi/StreamDecorator.java]: I used it at first, but then I took it out because I didn't know which features qualify for `@Incubating`. Now, it's clear.

File: hibernate-core/src/main/java/org/hibernate/query/spi/StreamDecorator.java (hunk 1 above)
Comment (vladmihalcea) [reply hibernate-core/src/main/java/org/hibernate/query/spi/StreamDecorator.java]: @sebersole @NathanQingyangXu I thought of that too, but then I run the tests with Java 13 and there was no issue. I've been using this approach in {FlexyPool](https://github.com/vladmihalcea/flexy-pool) as an alternative to dynamic proxies for the JDBC `Connection`. If there's a method added in a new JDK version, and we need to compile with JDK 8, we can just add the new methods without `@Override` and use Reflection to invoke the method on the delegate.

Another approach is to use Byte Buddy, but there will be many if/else checks to determine which methods we need to call `close` and which require creating a new `StreamDecorator`.

File: hibernate-core/src/test/java/org/hibernate/test/stream/basic/JpaStreamTest.java (hunk 2 above)
Comment (vladmihalcea) [reply hibernate-core/src/test/java/org/hibernate/test/stream/basic/JpaStreamTest.java]: Yes, we should totally add that test case. Luckily, the `AsbtractScrollableResults` returns immediately from `close` if the RS was already closed:

```
if ( this.closed ) {
	// noop if already closed
	return;
}
```

Comment (NathanQingyangXu) [reply hibernate-core/src/test/java/org/hibernate/test/stream/basic/JpaStreamTest.java]: Yeah, I noticed that, but adding testing case would make us sleep well if the above code was refactored.

File: hibernate-core/src/main/java/org/hibernate/query/spi/StreamDecorator.java (hunk 1 above)
Comment (vladmihalcea) [reply hibernate-core/src/main/java/org/hibernate/query/spi/StreamDecorator.java]: @sebersole @NathanQingyangXu Cool. I found why it works on Java 13. It looks like the JDK folks decided to mark the new `Stream` methods using `default`, so that's why the test run. However, we still need to add those so that we can either decorate the `Stream` or call `close` if it's a terminator operation. I'll issue a new commit based on these comments.

**Comment by NathanQingyangXu** [comment]
Do we have the concern that this feature will break some existing client code which assumes it is their responsibility to close the stream whenever they want? Seems not.
It might be a good idea to put some thing in our wonderful user guide to explain this cool feature to avoid unnecessary bug reporting or suchlike.

**Comment by vladmihalcea** [comment]
@NathanQingyangXu I added a test case for the close method and it works just fine. I think that in case this feature gets merged, it's a good idea to update the User Guide to explain how the JPA 2.2 Stream results work in Hibernate, and why it's a good idea to call a terminal operation sooner than later.

**Comment by vladmihalcea** [comment]
@NathanQingyangXu @sebersole I pushed the new commit based on our discussion.

I'm now thinking of adding a new test case that will run on Java 9+ only so that I can verify the outcome of the `takeWhile` and `dropWhile` methods. I want to make sure the `Stream` is closed for those methods as well when executing the terminal operation.

**Comment by sebersole** [comment]
There are some check-style failures.  Other than that, +1

**Comment by vladmihalcea** [comment]
I run the `check` phase to fix those issues as well.

**Comment by vladmihalcea** [comment]
I also added test cases for `takeWhile` and `doWhile`.

**Comment by vladmihalcea** [comment]
I updated the User Guide as well. I'll rebase and merge it tomorrow.

**Comment by vladmihalcea** [comment]
@NathanQingyangXu I rebased the branch from master and squashed all previous commits. If you can merge it, I'd really appreciate it. Thanks.

**Comment by NathanQingyangXu** [comment]
> @NathanQingyangXu I rebased the branch from master and squashed all previous commits. If you can merge it, I'd really appreciate it. Thanks.

I have no such rights for I am a community contributor only. But I guess Steve will do that soon.

**Comment by NathanQingyangXu** [comment]
> I updated the User Guide as well. I'll rebase and merge it tomorrow.

Thanks for the User Guide update. I used to finish a big PR to improve on it: https://github.com/hibernate/hibernate-orm/pull/3174. 
One thing I really love it is the testing code and easy verification. That is brilliant.
Seems lots of work needs to be done after v6 is released.

File: documentation/src/main/asciidoc/userguide/chapters/query/hql/HQL.adoc [hunk 4]
```diff
@@ -414,6 +424,29 @@ include::{sourcedir}/HQLTest.java[tags=hql-api-stream-example]
...
 
+[[jpql-api-stream]]
+==== Query streaming
+
+Since version 2.2, the JPA `Query` interface offers support for returning a `Stream` via the `getResultStream` method.
+
+Just like the `scroll` method, you can use a try-with-resources block to close the `Stream`
+prior to closing the currently running Persistence Context.
+
+Since Hibernate 5.4, the `Stream` is also closed when calling a terminal operation,
```
Comment (NathanQingyangXu) [inline documentation/src/main/asciidoc/userguide/chapters/query/hql/HQL.adoc]: technically speaking, this feature will appear in next version, i.e. 5.4.13, not ever since 5.4 was released.

File: documentation/src/test/java/org/hibernate/userguide/hql/HQLTest.java [hunk 5]
```diff
@@ -673,6 +671,38 @@ public void test_jpql_api_list_example() {
...
+			// tag::jpql-api-stream-terminal-operation[]
+			List<Person> persons = entityManager.createQuery(
+					"select p " +
+					"from Person p " +
+					"where p.name like :name", Person.class )
+			.setParameter( "name", "J%" )
+			.getResultStream()
+			.skip( 5 )
+			.limit( 5 )
+			.collect( Collectors.toList() );
```
Comment (NathanQingyangXu) [inline documentation/src/test/java/org/hibernate/userguide/hql/HQLTest.java]: It would be great to make consistent usage of code style. I even created a ticket to fix the various format issue (sometimes indentation produces really bad result because of mixed usage of TAB and space, e.g.).
Reformatting our testing code with the https://github.com/hibernate/hibernate-ide-codestyles will solve the issue easily.
So I am okay with this (there is some minor format issue above like Line 678. There should be a space after `try` and before `{`).
Many thanks for updating the User Guide, however.
